flowmachine/flowmachine/features/subscriber/contact_modal_location_distance.py

```diff
@@ -0,0 +1,189 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+# -*- coding: utf-8 -*-
+"""
+Statistics for the distance between subscriber's own modal
+location and its contacts' modal location.
+"""
+
+from .metaclasses import SubscriberFeature
+
+from .modal_location import ModalLocation
+from .contact_balance import ContactBalance
+from ..spatial.distance_matrix import DistanceMatrix
+from .daily_location import daily_location
+
+from flowmachine.utils import get_columns_for_level, list_of_dates
+
+valid_stats = {"count", "sum", "avg", "max", "min", "median", "stddev", "variance"}
+
+
+class ContactModalLocationDistance(SubscriberFeature):
+    """
+    This class calculates statistics for the distance between subscriber's own modal
+    location and its contacts' modal location.
+
+    Parameters
+    ----------
+    start, stop : str
+         iso-format start and stop datetimes
+    statistic : {'count', 'sum', 'avg', 'max', 'min', 'median', 'mode', 'stddev', 'variance'}, default 'sum'
+        Defaults to sum, aggregation statistic over the durations.
+    method : str, default 'last'
+        The method by which to calculate the location of the subscriber.
+        This can be either 'most-common' or last. 'most-common' is
+        simply the modal location of the subscribers, whereas 'lsat' is
+        the location of the subscriber at the time of the final call in
+        the data.
+    contact_balance: features.subscriber.ContactBalance, default None.
+        An instance of ContactBalance. If an instance is not provide, the
+        ContactBalance is instantiated with the same parameters as for the
+        features.subscriber.ModalLocation
+    hours : 2-tuple of floats, default 'all'
+        Restrict the analysis to only a certain set
+        of hours within each day.
+    subscriber_subset : str, list, flowmachine.core.Query, flowmachine.core.Table, default None
+        If provided, string or list of string which are msisdn or imeis to limit
+        results to; or, a query or table which has a column with a name matching
+        subscriber_identifier (typically, msisdn), to limit results to.
+    direction : {'in', 'out', 'both'}, default 'out'
+        Whether to consider calls made, received, or both. Defaults to 'out'.
+    column_name : str
+        Optionally specify a non-default column name. Required if level is 'polygon'.
+
+    Example
+    -------
+
+    >>> s = ContactModalLocationDistance("2016-01-01", "2016-01-03", statistic="avg")
+    >>> s.get_dataframe()
+
+        subscriber           value
+    gwAynWXp4eWvxGP7    298.721500
+    GnyZMedmKQ4X78Wa    290.397556
+    BKMy1nYEZpnoEA7G     78.919136
+    m4L326vrwE6elJxQ    249.033988
+    NG1km5NzBg5JD8nj    188.679378
+               ...           ...
+    """
+
+    def __init__(
```

Comment:
I'm wondering if it wouldn't be preferable to have this take either or both of the ML and CB as args and either create one from the other, or check they're compatible. I think this way kind of gives the illusion of being more flexible, but actually isn't because you've got lots of args you can set but they're the same across the two subqueries. Would also allow you to calculate this for things other than a ML e.g. a Hartigan clustered location, or a daily location.

Reply:
I had initially created the class in the way you described above. But then I reverted to what I have now because I wanted the signature of the feature classes to be somewhat consistent.

Requesting the ML and CB classes actually make the whole thing much easier to define. In which case, it would also make sense to allow for other classes instead of ML such as Hartigan and/or daily location. 

One reason why I wanted to exclude Hartigan was because of what I discuss in the next comment below.

I would then also rename the class to something more broad like `ContactHomeDistance`?